### PR TITLE
Adds validation that tablet does not have operation id and location at the same time

### LINKED ISSUE #3443: Add validation that tablet does not have operation id and location at the same time
Accumulo has validation that a tablet does not have a current and future location in the TabletManagementIterator.  Need to add additional validation that a tablet does not have an operation id and location at the same time.

When a tablet has an operation id set it means a destructive operation like merge,split, delete is underway and its not expected that the tablet would be assigned/hosted.  Therefore if a tablet with a location and operation id is ever seen its an indication of a problem that should be logged.

### PR DESCRIPTION
**Pull request by DomGarguilo**
Closes #3443

### CODE REVIEW

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java [hunk 1]
```diff
@@ -182,6 +183,31 @@ public void testFutureAndCurrent() {
...
+    mutation.at().family(MetadataSchema.TabletsSection.ServerColumnFamily.STR_NAME)
+        .qualifier(OPID_QUAL).put(TabletOperationType.DELETING + ":" + formatTid(tableId));
+    mutation.at().family(FutureLocationColumnFamily.NAME).qualifier("s001").put("server1:8555");
+
+    SortedMap<Key,Value> rowMap = toRowMap(mutation);
+
+    assertThrows(IllegalStateException.class,
+        () -> TabletMetadata.convertRow(rowMap.entrySet().iterator(),
+            EnumSet.allOf(ColumnType.class), false, false),
+        "Should not be able to have both a future location and operation id set");
```
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java:204]: This test message is not correct.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 2]
```diff
@@ -597,6 +602,18 @@ private void setLocationOnce(String val, String qual, LocationType lt, boolean s
     location = new Location(val, qual, lt);
   }
 
+  private void setOperationIdOnce(String val, boolean suppressError) {
+    if (location != null) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:606]: Could add check that operation Id is null also.

```suggestion
    if (location != null || operationId != null) {
```

Could add a similar check of the operationId in the earlier method `setLocationOnce()`.  May need to adjust the erros messages in each method.

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java (hunk 1 above)
Comment (DomGarguilo) [reply core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java]: Fixed in 3fdcb2c

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 2 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java]: `operationId` is not used in `setLocationOnce()`. Should we still check that it is not null there? Also, are you suggesting that the existing error messages be changed if the `operationId != null` is added? If so how might they be changed?

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 3]
```diff
@@ -597,6 +602,18 @@ private void setLocationOnce(String val, String qual, LocationType lt, boolean s
     location = new Location(val, qual, lt);
   }
 
+  private void setOperationIdOnce(String val, boolean suppressError) {
+    if (location != null || operationId != null) {
```
Comment (ddanielr) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:606]: Is it still an error condition if an operationId is already set and location is null?

**Comment by DomGarguilo** [comment]
After adding the new boolean to the toString method I looked around and found what seems like an improvement to the method using ToStringBuilder, which I hadn't seen before and it seems pretty nice. The output before and after a2814a0 should be identical.  Its not exactly related to this PR but I thought I would include it here while I'm editing the method anyways.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 2 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java]: The operation id and location are mutually exclusive, we should never see a tablet with both set.  That is why I was thinking of adding something to the set location method.  When we set a location, its expected that the operation id is null.  When we set an operation id, its expected that the location is nuill.

Was thinknig th error message could include the location and operationid if both are being checked.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 3 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:612]: The way the operation id is stored in the metadata table, under normal circumstances we would never expect to see it twice for a tablet.  When we set the operation id it expected that the location and operation id are null, anything else is unexpected.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 4]
```diff
@@ -443,6 +450,10 @@ public boolean isFutureAndCurrentLocationSet() {
     return futureAndCurrentLocationSet;
   }
 
+  public boolean isOperationIdAndCurrentLocationSet() {
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:453]: Check `TabletManagementIterator.computeTabletManagementActions`and [TabletGroupWatcher](https://github.com/apache/accumulo/blob/elasticity/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L252) to see how to wire this up.

Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:453]: Added in e79f423

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 2 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:644]: I tried to fix this in 84e15bd

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 5]
```diff
@@ -594,9 +614,44 @@ private void setLocationOnce(String val, String qual, LocationType lt, boolean s
...
+   * @throws IllegalStateException if an operation id or location is already set
+   */
+  private void setOperationIdOnce(String val, boolean suppressError) {
+    // make sure there is not already an operation ID
+    if (operationId != null) {
+      if (!suppressError) {
+        throw new IllegalStateException(
+            "Attempted to set operation id for tablet that already has one. table ID: " + tableId
+                + " endrow: " + endRow + " -- operation id: " + operationId);
+      }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:641]: Seems like the boolean should be set here.

```suggestion
      }
      operationIdAndCurrentLocationSet = true;
```

**Review by keith-turner — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 5 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:641]: Should it be set here in addition to where it is already set (within the if block below this)? Or should it only be set here?

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 6]
```diff
@@ -594,9 +614,44 @@ private void setLocationOnce(String val, String qual, LocationType lt, boolean s
       }
       futureAndCurrentLocationSet = true;
     }
+    if (operationId != null) {
+      if (!suppressError) {
+        throw new IllegalStateException(
+            "Attempted to set location for tablet with an operation id. table ID: " + tableId
+                + " endrow: " + endRow + " -- operation id: " + operationId);
+      }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:622]:
```suggestion
      }
      operationIdAndCurrentLocationSet = true;
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 7]
```diff
@@ -594,9 +614,44 @@ private void setLocationOnce(String val, String qual, LocationType lt, boolean s
...
+   */
+  private void setOperationIdOnce(String val, boolean suppressError) {
+    // make sure there is not already an operation ID
+    if (operationId != null) {
+      if (!suppressError) {
+        throw new IllegalStateException(
+            "Attempted to set operation id for tablet that already has one. table ID: " + tableId
+                + " endrow: " + endRow + " -- operation id: " + operationId);
+      }
+    }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:642]: Thinking about this a bit more, it shoul dnot be possible to set the operation id twice with the current metadata schema.  So it being set twice is really really unexpected, so lets not have any handling for that and just always throw an exception.

```suggestion
    Preconditions.checkState(operationId == null);
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 5 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:641]: ignore this comment I made another comment that supersedes this one.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 7 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java]: Added in e13003c

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 6 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:622]: Added in e13003c

**Review by keith-turner — APPROVED** [review]

**Review by Manno15 — APPROVED** [review]

**Pull request merged**